internal/db/resource.go

```diff
@@ -106,3 +104,68 @@ func (d *Database) ListResources(ctx context.Context, limit, offset int) (_ []Re
 	}
 	return resources, nil
 }
+
+// buildQuery is an utility function to build the database query according to two optional parameters.
+// nameFilter: used to match resources name.
+// typeFilter used to match resources type. If this is not empty the resources are fetched from a single table.
+func buildQuery(db *gorm.DB, offset, limit int, nameFilter, typeFilter string) (*gorm.DB, error) {
+	query := `
+	? UNION ? UNION ? UNION ? UNION ?
+	ORDER BY type, id
+	OFFSET ?
+	LIMIT  ?;
+	`
+	applicationOffersQuery := db.Select(SELECT_APPLICATION_OFFERS).
+		Model(&dbmodel.ApplicationOffer{}).
+		Joins("JOIN models ON application_offers.model_id = models.id")
```

Comment:
I'd have the `JOIN`s already set in the const queries. Basically we don't get type-safe benefits from using GORM here (except for type mapping of the result).

Reply:
so you are suggesting to get rid of gorm all togheter and using raw queries?
i think using gorm + the where clause as you suggested you suggested it's clearer!

Reply:
but i do agree that gorm here doesn't provide type safety but just a tidier organization than manipulating SQL strings

Reply:
@babakks i tried using the RAW query for each table but the code was more confusing.
I think this iteration provides clarity and ease of manipulation.